File: platforms/android/stato/src/main/java/com/facebook/stato/core/StatoObject.java
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */
package com.facebook.stato.core;

import java.util.Arrays;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class StatoObject {
  final JSONObject mJson;

  public StatoObject(JSONObject json) {
    mJson = (json != null ? json : new JSONObject());
  }

  public StatoObject(String json) {
    try {
      mJson = new JSONObject(json);
    } catch (JSONException e) {
      throw new RuntimeException(e);
    }
  }

  public StatoDynamic getDynamic(String name) {
    return new StatoDynamic(mJson.opt(name));
  }

  public String getString(String name) {
    if (mJson.isNull(name)) {
      return null;
    }
    return mJson.optString(name);
  }

  public int getInt(String name) {
    return mJson.optInt(name);
  }

  public long getLong(String name) {
    return mJson.optLong(name);
  }

  public float getFloat(String name) {
    return (float) mJson.optDouble(name);
  }

  public double getDouble(String name) {
    return mJson.optDouble(name);
  }

  public boolean getBoolean(String name) {
    return mJson.optBoolean(name);
  }

  public StatoObject getObject(String name) {
    final Object o = mJson.opt(name);
    return new StatoObject((JSONObject) o);
  }

  public StatoArray getArray(String name) {
    final Object o = mJson.opt(name);
    return new StatoArray((JSONArray) o);
  }

  public boolean contains(String name) {
    return mJson.has(name);
  }

  public String toJsonString() {
    return toString();
  }

  @Override
  public String toString() {
    return mJson.toString();
  }

  @Override
  public boolean equals(Object o) {
    return mJson.toString().equals(o.toString());
  }

  @Override
  public int hashCode() {
    return mJson.hashCode();
  }

  public static class Builder {
    private final JSONObject mJson;

    public Builder() {
      mJson = new JSONObject();
    }

    public Builder put(String name, Object obj) {
      if (obj == null) {
        return put(name, (String) null);
      } else if (obj instanceof Integer) {
        return put(name, (Integer) obj);
      } else if (obj instanceof Long) {
        return put(name, (Long) obj);
      } else if (obj instanceof Float) {
        return put(name, (Float) obj);
      } else if (obj instanceof Double) {
        return put(name, (Double) obj);
      } else if (obj instanceof String) {
        return put(name, (String) obj);
      } else if (obj instanceof Boolean) {
        return put(name, (Boolean) obj);
      } else if (obj instanceof Object[]) {
        return put(name, Arrays.deepToString((Object[]) obj));
      } else if (obj instanceof StatoObject) {
        return put(name, (StatoObject) obj);
      } else if (obj instanceof StatoObject.Builder) {
        return put(name, (StatoObject.Builder) obj);
      } else if (obj instanceof StatoArray) {
        return put(name, (StatoArray) obj);
      } else if (obj instanceof StatoArray.Builder) {
        return put(name, (StatoArray.Builder) obj);
      } else if (obj instanceof StatoValue) {
        return put(name, ((StatoValue) obj).toStatoObject());
      } else {
        return put(name, obj.toString());
      }
    }

    public Builder put(String name, String s) {
      try {
        mJson.put(name, s);
      } catch (JSONException e) {
        throw new RuntimeException(e);
      }
      return this;
    }

    public Builder put(String name, Integer i) {
      try {
        mJson.put(name, i);
      } catch (JSONException e) {
        throw new RuntimeException(e);
      }
      return this;
    }

    public Builder put(String name, Long l) {
      try {
        mJson.put(name, l);
      } catch (JSONException e) {
        throw new RuntimeException(e);
      }
      return this;
    }

    public Builder put(String name, Float f) {
      try {
        mJson.put(name, Float.isNaN(f) ? null : f);
      } catch (JSONException e) {
        throw new RuntimeException(e);
      }
      return this;
    }

    public Builder put(String name, Double d) {
      try {
        mJson.put(name, Double.isNaN(d) ? null : d);
      } catch (JSONException e) {
        throw new RuntimeException(e);
      }
      return this;
    }

    public Builder put(String name, Boolean b) {
      try {
        mJson.put(name, b);
      } catch (JSONException e) {
        throw new RuntimeException(e);
      }
      return this;
    }

    public Builder put(String name, StatoValue v) {
      return put(name, v.toStatoObject());
    }

    public Builder put(String name, StatoArray a) {
      try {
        mJson.put(name, a == null ? null : a.mJson);
      } catch (JSONException e) {
        throw new RuntimeException(e);
      }
      return this;
    }

    public Builder put(String name, StatoArray.Builder b) {
      return put(name, b.build());
    }

    public Builder put(String name, StatoObject o) {
      try {
        mJson.put(name, o == null ? null : o.mJson);
      } catch (JSONException e) {
        throw new RuntimeException(e);
      }
      return this;
    }

    public Builder put(String name, StatoObject.Builder b) {
      return put(name, b.build());
    }

    public StatoObject build() {
      return new StatoObject(mJson);
    }
  }
}

File: platforms/android/android-stato/src/main/kotlin/org/stato/core/StatoObject.kt
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */
package org.stato.core

import java.util.Arrays
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class StatoObject(json: JSONObject? = null) {
  internal val json: JSONObject = json ?: JSONObject()

  constructor(json: String) : this(JSONObject(json))

  fun getDynamic(name: String): StatoDynamic {
    return StatoDynamic(json.opt(name))
  }

  fun getString(name: String): String? {
    return if (json.isNull(name)) {
      null
    } else json.optString(name)
  }

  fun getInt(name: String): Int {
    return json.optInt(name)
  }

  fun getLong(name: String): Long {
    return json.optLong(name)
  }

  fun getFloat(name: String): Double {
    return json.optDouble(name)
  }

  fun getDouble(name: String): Double {
    return json.optDouble(name)
  }

  fun getBoolean(name: String): Boolean {
    return json.optBoolean(name)
  }

  fun getObject(name: String): StatoObject {
    val o = json.opt(name)
    return StatoObject(o as JSONObject)
  }

  fun getArray(name: String): StatoArray {
    val o = json.opt(name)
    return StatoArray(o as JSONArray)
  }

  operator fun contains(name: String): Boolean {
    return json.has(name)
  }

  fun toJsonString(): String {
    return toString()
  }


  override fun toString(): String {
    return json.toString()
  }

    override fun equals(other: Any?): Boolean {
    return json.toString() == other?.toString()
  }

    override fun hashCode(): Int {
    return json.hashCode()
  }

  class Builder {
    private val json: JSONObject = JSONObject()

    fun put(name: String): Builder {
      json.put(name, JSONObject.NULL)
      return this
    }

    fun put(name: String, obj: Any?): Builder {
      return when (obj) {
        null -> put(name)
        is Int -> put(name, obj )
        is Long -> put(name, obj )
        is Float -> put(name, obj )
        is Double -> put(name, obj)
        is String -> put(name, obj )
        is Boolean -> put(name, obj )
        is Array<*> -> put(name, Arrays.deepToString(obj))
        is StatoObject -> put(name, obj)
        is Builder -> put(name, obj)
        is StatoArray -> put(name, obj)
        is StatoArray.Builder -> put(name, obj)
        is StatoValue -> put(name, (obj).toStatoObject())
        else -> put(name, obj.toString())
      }
    }

    fun put(name: String, s: String?): Builder {
      try {
        json.put(name, s ?: JSONObject.NULL)
      } catch (e: JSONException) {
        throw RuntimeException(e)
      }

      return this
    }

    fun put(name: String, i: Int?): Builder {
      try {
        json.put(name, i ?: JSONObject.NULL)
      } catch (e: JSONException) {
        throw RuntimeException(e)
      }

      return this
    }

    fun put(name: String, l: Long?): Builder {
      try {
        json.put(name, l ?: JSONObject.NULL)
      } catch (e: JSONException) {
        throw RuntimeException(e)
      }

      return this
    }

    fun put(name: String, f: Float?): Builder {
      try {
        json.put(name, (if (Float.NaN == f) JSONObject.NULL else f) ?: JSONObject.NULL)
      } catch (e: JSONException) {
        json.put(name, JSONObject.NULL)
        //throw RuntimeException(e)
      }

      return this
    }

    fun put(name: String, d: Double?): Builder {
      try {
        json.put(name, (if (Double.NaN == d) null else d) ?: JSONObject.NULL)
      } catch (e: JSONException) {
        throw RuntimeException(e)
      }

      return this
    }

    fun put(name: String, b: Boolean?): Builder {
      try {
        json.put(name, b?: JSONObject.NULL)
      } catch (e: JSONException) {
        throw RuntimeException(e)
      }

      return this
    }

    fun put(name: String, v: StatoValue?): Builder {
      return put(name, v?.toStatoObject() ?: JSONObject.NULL)
    }

    fun put(name: String, a: StatoArray?): Builder {
      try {
        json.put(name, if (a == null) null else a!!.json)
      } catch (e: JSONException) {
        throw RuntimeException(e)
      }

      return this
    }

    fun put(name: String, b: StatoArray.Builder?): Builder {
      return put(name, b?.build() ?: JSONObject.NULL)
    }

    fun put(name: String, o: StatoObject?): Builder {
      try {
        json.put(name, o?.json ?: JSONObject.NULL)
      } catch (e: JSONException) {
        throw RuntimeException(e)
      }

      return this
    }

    fun put(name: String, b: Builder): Builder {
      return put(name, b.build())
    }

    fun build(): StatoObject {
      return StatoObject(json)
    }
  }
}
